### PR TITLE
Iox #842 cleanup memory provider and memory block

### LINKED ISSUE #842: Cleanup MemoryProvider and MemoryBlock
## Brief feature description

There are some minor issues with the `MemoryProvider` and `MemoryBlock`

## Detailed information

- [x] use `@copydoc Base::method` instead of `@brief Implementation of Base::method`
- [x] virtual dtor for base classes
- [x] make `MemoryBlock::memoryAvailable` and `MemoryBlock::destroy` protected
- [x] use `cxx::not_null` for parameter of `MemoryBlock::memoryAvailable`

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

## Notes for Reviewer

This fixes some of the requests from #840 

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #842

### CODE REVIEW

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/memory/mempool_collection_memory_block.hpp [hunk 1]
```diff
@@ -45,27 +45,27 @@ class MemPoolCollectionMemoryBlock final : public MemoryBlock
     MemPoolCollectionMemoryBlock& operator=(const MemPoolCollectionMemoryBlock&) = delete;
     MemPoolCollectionMemoryBlock& operator=(MemPoolCollectionMemoryBlock&&) = delete;
 
-    /// @brief Implementation of MemoryBlock::size
-    /// @return the size of type T
+    /// @copydoc MemoryBlock::size
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/memory/mempool_collection_memory_block.hpp:48]: Please adapt the copyright header

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/memory/mempool_segment_manager_memory_block.hpp [hunk 2]
```diff
@@ -40,27 +40,27 @@ class MemPoolSegmentManagerMemoryBlock : public MemoryBlock
     MemPoolSegmentManagerMemoryBlock& operator=(const MemPoolSegmentManagerMemoryBlock&) = delete;
     MemPoolSegmentManagerMemoryBlock& operator=(MemPoolSegmentManagerMemoryBlock&&) = delete;
 
-    /// @brief Implementation of MemoryBlock::size
-    /// @return the size of for SegmentManager
+    /// @copydoc MemoryBlock::size
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/memory/mempool_segment_manager_memory_block.hpp:43]: Please adapt the copyright header

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/memory/port_pool_memory_block.hpp [hunk 3]
```diff
@@ -40,27 +40,27 @@ class PortPoolMemoryBlock : public MemoryBlock
     PortPoolMemoryBlock& operator=(const PortPoolMemoryBlock&) = delete;
     PortPoolMemoryBlock& operator=(PortPoolMemoryBlock&&) = delete;
 
-    /// @brief Implementation of MemoryBlock::size
-    /// @return the size of for all the ports
+    /// @copydoc MemoryBlock::size
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/memory/port_pool_memory_block.hpp:43]: Copyright header update needed here too ;)

File: iceoryx_posh/include/iceoryx_posh/roudi/memory/generic_memory_block.hpp [hunk 4]
```diff
@@ -41,17 +41,14 @@ class GenericMemoryBlock final : public MemoryBlock
     GenericMemoryBlock& operator=(const GenericMemoryBlock&) = delete;
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/roudi/memory/generic_memory_block.hpp:41]: Same here

File: iceoryx_posh/include/iceoryx_posh/roudi/memory/memory_block.hpp [hunk 5]
```diff
@@ -34,6 +35,7 @@ class MemoryBlock
 
   public:
     MemoryBlock() noexcept = default;
+    virtual ~MemoryBlock() = default;
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/roudi/memory/memory_block.hpp:38]: `noexcept`?

**Review by mossmaurice — CHANGES_REQUESTED** [review]
Please double check all the copyright headers, otherwise LGTM.

In general, should there be death tests added for methods that use `not_null`?

**Comment by elBoberido** [comment]
Regarding the death test for `not_null`. In this specific case it's not really possible since it's a protected method and on the caller side this is only called if it's not a `nullptr`. I would have used a reference, but that's not possible with a `void`.

In general I also wouldn't do death tests for parameter with `not_null` since we also don't do death tests for methods with references as parameter and you could easily have something like this
```cpp
int foo (int& i) {
    return i;
}
int bar(int* i) {
    return foo(*i);
}
int main() {
    return bar(nullptr); // crash will happen on `return i;` not `return foo(*i);`
}
```
As long as we already tested that `not_null` itself terminates, it is not necessary to test it again everywhere else it is used.

File: iceoryx_posh/include/iceoryx_posh/roudi/memory/memory_block.hpp [hunk 6]
```diff
@@ -52,19 +55,20 @@ class MemoryBlock
...
     /// @brief This function is called once the memory is available and is therefore the earliest possibility to use the
     /// memory.
     /// @param [in] memory pointer to a valid memory block, the same one that the memory() member function would return
-    virtual void memoryAvailable(void* memory) noexcept;
-
-    /// @brief This function provides the pointer to the requested memory.
-    /// @return an optional pointer to a memory block with the requested size and alignment if the memory is available,
-    /// otherwise a cxx::nullopt_t
-    cxx::optional<void*> memory() const noexcept;
+    virtual void memoryAvailable(cxx::not_null<void*> memory) noexcept;
```
Comment (dkroenke) [inline iceoryx_posh/include/iceoryx_posh/roudi/memory/memory_block.hpp:71]: I understand the purpose of this "callback" but find the name a bit misleading.
It is called automatically at a certain condition so we could set the prefix "on" here.
Maybe something like "onMemoryBlockCreated". But these are only my 2 cents.

**Review by mossmaurice — DISMISSED** [review]

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/memory/mempool_segment_manager_memory_block.hpp [hunk 7]
```diff
@@ -40,27 +41,27 @@ class MemPoolSegmentManagerMemoryBlock : public MemoryBlock
...
-    /// @brief Implementation of MemoryBlock::destroy
-    /// This will clean up the SegmentManager
-    void destroy() noexcept override;
-
     /// @brief This function enables the access to the SegmentManager
     /// @return an optional pointer to the underlying type, cxx::nullopt_t if value is not initialized
     cxx::optional<mepoo::SegmentManager<>*> segmentManager() const noexcept;
 
+  protected:
+    /// @copydoc MemoryBlock::memoryAvailable
```
Comment (dkroenke) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/memory/mempool_segment_manager_memory_block.hpp:57]: should be: onMemoryAvailable

**Review by mossmaurice — APPROVED** [review]

**Review by dkroenke — APPROVED** [review]

**Pull request merged**